File: stripe/src/test/java/com/stripe/android/model/ConfirmSetupIntentParamsTest.java
package com.stripe.android.model;

import org.junit.Test;

import java.util.Map;
import java.util.Objects;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertTrue;

public class ConfirmSetupIntentParamsTest {

    @Test
    public void shouldUseStripeSdk() {
        final ConfirmSetupIntentParams confirmSetupIntentParams =
                ConfirmSetupIntentParams.create(
                        "pm_123", "client_secret", "return_url");
        assertFalse(confirmSetupIntentParams.shouldUseStripeSdk());

        assertTrue(confirmSetupIntentParams
                .withShouldUseStripeSdk(true)
                .shouldUseStripeSdk());
    }

    @Test
    public void create_withPaymentMethodId_shouldPopulateParamMapCorrectly() {
        final ConfirmSetupIntentParams confirmSetupIntentParams =
                ConfirmSetupIntentParams.create(
                        "pm_12345",
                        "client_secret",
                        null
                );
        final Map<String, Object> params = confirmSetupIntentParams.toParamMap();
        assertNull(params.get(ConfirmStripeIntentParams.API_PARAM_PAYMENT_METHOD_DATA));
        assertEquals("pm_12345",
                params.get(ConfirmStripeIntentParams.API_PARAM_PAYMENT_METHOD_ID));
    }

    @SuppressWarnings("unchecked")
    @Test
    public void create_withPaymentMethodCreateParams_shouldPopulateParamMapCorrectly() {
        final PaymentMethodCreateParams.Card expectedCard =
                new PaymentMethodCreateParams.Card.Builder()
                        .setNumber("4242424242424242")
                        .setCvc("123")
                        .setExpiryMonth(8)
                        .setExpiryYear(2019)
                        .build();

        final ConfirmSetupIntentParams confirmSetupIntentParams =
                ConfirmSetupIntentParams.create(
                        PaymentMethodCreateParams.create(expectedCard, null),
                        "client_secret",
                        null
                );
        final Map<String, Object> params = confirmSetupIntentParams.toParamMap();
        assertNull(params.get(ConfirmStripeIntentParams.API_PARAM_PAYMENT_METHOD_ID));
        final Map<String, Object> paymentMethodData = (Map<String, Object>)
                Objects.requireNonNull(
                        params.get(ConfirmStripeIntentParams.API_PARAM_PAYMENT_METHOD_DATA));
        assertEquals("card", paymentMethodData.get("type"));
        assertNotNull(paymentMethodData.get("card"));
    }

    @Test
    public void create_withShouldUseStripeSdk_shouldKeepPaymentMethodData() {
        final PaymentMethodCreateParams.Card expectedCard =
                new PaymentMethodCreateParams.Card.Builder()
                        .setNumber("4242424242424242")
                        .setCvc("123")
                        .setExpiryMonth(8)
                        .setExpiryYear(2019)
                        .build();

        ConfirmSetupIntentParams confirmSetupIntentParams =
                ConfirmSetupIntentParams.create(
                        PaymentMethodCreateParams.create(expectedCard, null),
                        "client_secret",
                        null
                );

        assertEquals(false, confirmSetupIntentParams.shouldUseStripeSdk());
        confirmSetupIntentParams = confirmSetupIntentParams.withShouldUseStripeSdk(true);
        assertEquals(true, confirmSetupIntentParams.shouldUseStripeSdk());


        // Ensures that payment method data is still present after the withShouldUseStripeSdk call
        final Map<String, Object> params = confirmSetupIntentParams.toParamMap();
        assertNull(params.get(ConfirmStripeIntentParams.API_PARAM_PAYMENT_METHOD_ID));
        final Map<String, Object> paymentMethodData = (Map<String, Object>)
                Objects.requireNonNull(
                        params.get(ConfirmStripeIntentParams.API_PARAM_PAYMENT_METHOD_DATA));
        assertEquals("card", paymentMethodData.get("type"));
        assertNotNull(paymentMethodData.get("card"));
    }
}

File: stripe/src/test/java/com/stripe/android/model/ConfirmSetupIntentParamsTest.kt
package com.stripe.android.model

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue

class ConfirmSetupIntentParamsTest {

    @Test
    fun shouldUseStripeSdk_withPaymentMethodId() {
        val confirmSetupIntentParams = ConfirmSetupIntentParams.create(
            "pm_123", "client_secret", "return_url")
        assertFalse(confirmSetupIntentParams.shouldUseStripeSdk())

        assertTrue(confirmSetupIntentParams
            .withShouldUseStripeSdk(true)
            .shouldUseStripeSdk())
    }

    @Test
    fun shouldUseStripeSdk_withPaymentMethodCreateParams() {
        val confirmSetupIntentParams = ConfirmSetupIntentParams.create(
            PaymentMethodCreateParamsFixtures.DEFAULT_CARD,
            "client_secret",
            "return_url"
        )
        assertFalse(confirmSetupIntentParams.shouldUseStripeSdk())

        assertTrue(confirmSetupIntentParams
            .withShouldUseStripeSdk(true)
            .shouldUseStripeSdk())
    }

    @Test
    fun create_withPaymentMethodId_shouldPopulateParamMapCorrectly() {
        val confirmSetupIntentParams = ConfirmSetupIntentParams.create(
            "pm_12345",
            "client_secret",
            null
        )
        val params = confirmSetupIntentParams.toParamMap()
        assertNull(params[ConfirmStripeIntentParams.PARAM_PAYMENT_METHOD_DATA])
        assertEquals("pm_12345",
            params[ConfirmStripeIntentParams.PARAM_PAYMENT_METHOD_ID])
    }

    @Test
    fun create_withPaymentMethodCreateParams_shouldPopulateParamMapCorrectly() {
        val confirmSetupIntentParams = ConfirmSetupIntentParams.create(
            PaymentMethodCreateParamsFixtures.DEFAULT_CARD,
            "client_secret", null
        )
        val params = confirmSetupIntentParams.toParamMap()
        assertNull(params[ConfirmStripeIntentParams.PARAM_PAYMENT_METHOD_ID])
        val paymentMethodData =
            params[ConfirmStripeIntentParams.PARAM_PAYMENT_METHOD_DATA] as Map<String, Any>
        assertEquals("card", paymentMethodData["type"])
        assertNotNull(paymentMethodData["card"])
    }

    @Test
    fun create_withoutPaymentMethod() {
        val params = ConfirmSetupIntentParams.createWithoutPaymentMethod("client_secret")
        assertNull(params.paymentMethodCreateParams)
        assertNull(params.paymentMethodId)
    }

    @Test
    fun create_withSepaDebitPaymentMethodParams_shouldUseDefaultMandateDataIfNotSpecified() {
        val expectedParams = mapOf(
            "client_secret" to CLIENT_SECRET,
            "use_stripe_sdk" to false,
            "mandate_data" to mapOf(
                "customer_acceptance" to mapOf(
                    "type" to "online",
                    "online" to mapOf(
                        "infer_from_client" to true
                    )
                )
            ),
            "payment_method_data" to mapOf(
                "type" to "sepa_debit",
                "sepa_debit" to mapOf(
                    "iban" to "my_iban"
                )
            )
        )

        val actualParams = ConfirmSetupIntentParams.create(
            clientSecret = CLIENT_SECRET,
            paymentMethodCreateParams = PaymentMethodCreateParamsFixtures.DEFAULT_SEPA_DEBIT
        ).toParamMap()
        assertEquals(expectedParams, actualParams)
    }

    @Test
    fun create_withSepaDebitPaymentMethodParams_shouldUseMandateDataIfSpecified() {
        val expectedParams = mapOf(
            "client_secret" to CLIENT_SECRET,
            "use_stripe_sdk" to false,
            "mandate_data" to mapOf(
                "customer_acceptance" to mapOf(
                    "type" to "online",
                    "online" to mapOf(
                        "ip_address" to "127.0.0.1",
                        "user_agent" to "my_user_agent"
                    )
                )
            ),
            "payment_method_data" to mapOf(
                "type" to "sepa_debit",
                "sepa_debit" to mapOf(
                    "iban" to "my_iban"
                )
            )
        )

        val actualParams = ConfirmSetupIntentParams.create(
            clientSecret = CLIENT_SECRET,
            paymentMethodCreateParams = PaymentMethodCreateParamsFixtures.DEFAULT_SEPA_DEBIT,
            mandateData = MandateDataParamsFixtures.DEFAULT
        ).toParamMap()
        assertEquals(expectedParams, actualParams)
    }

    @Test
    fun create_withSepaDebitPaymentMethodParams_shouldUseMandateIdIfSpecified() {
        val expectedParams = mapOf(
            "client_secret" to CLIENT_SECRET,
            "use_stripe_sdk" to false,
            "mandate" to "mandate_123456789",
            "payment_method_data" to mapOf(
                "type" to "sepa_debit",
                "sepa_debit" to mapOf(
                    "iban" to "my_iban"
                )
            )
        )
        val actualParams =
            ConfirmSetupIntentParams.create(
                clientSecret = CLIENT_SECRET,
                paymentMethodCreateParams = PaymentMethodCreateParamsFixtures.DEFAULT_SEPA_DEBIT,
                mandateId = "mandate_123456789"
            ).toParamMap()
        assertEquals(expectedParams, actualParams)
    }

    private companion object {
        private const val CLIENT_SECRET = "seti_1CkiBMLENEVhOs7YMtUehLau_secret_sw1VaYPGZA"
    }
}
